### ISSUE TITLE
LoRA et QLoRA — fine-tuning à moindre coût

### ISSUE THREAD

**Issue by khalilbenaz** [khalilbenaz/khalilbenaz.github.io#134]
---
lede: LoRA (Low-Rank Adaptation) permet de fine-tuner un 7B sur un GPU consumer en quelques heures. QLoRA va plus loin en quantifiant le modèle de base.
lede_en: LoRA (Low-Rank Adaptation) lets you fine-tune a 7B on a consumer GPU in hours. QLoRA goes further by quantizing the base model.
title_en: LoRA and QLoRA — fine-tuning on a budget
---

## Le problème du fine-tuning classique

Fine-tuner un modèle entier (full fine-tuning) demande :

- Mémoire = taille du modèle × 4-6 (poids + gradients + optimizer state).
- Pour un 7B : ~70 GB de VRAM. Pour un 70B : ~500 GB.
- Coût : plusieurs milliers de dollars en compute par run.

Hors d'atteinte pour 99 % des équipes.

## L'idée de LoRA

Au lieu d'entraîner tous les poids, **LoRA ajoute de petites matrices d'adaptation** (rang faible) sur certaines couches (typiquement les couches d'attention). Ces matrices ont ~1 % des paramètres du modèle.

```
Modèle complet : W (7B params, frozen)
LoRA : W' = W + A·B  où A,B sont petits (~70M params trainables)
```

Conséquences :

- **Mémoire** : ~10 GB pour fine-tuner un 7B.
- **Temps** : quelques heures sur un GPU consumer (RTX 4090).
- **Output** : un fichier `lora.safetensors` de ~50 MB à appliquer au base model.

## QLoRA — encore plus efficient

QLoRA combine LoRA + quantization du base model en 4-bit. Résultat :

- **Fine-tuner un 70B sur un seul A100 80GB** (vs 8× A100 en full FT).
- **Fine-tuner un 7B sur RTX 3090 24GB**.
- Qualité comparable au full fine-tuning.

## Outils pratiques

- **PEFT** (Hugging Face) : la lib de référence Python.
- **Axolotl** : framework wrapper avec configs YAML lisibles.
- **MLX-LM** : pour les Mac M-series, fine-tuning natif.
- **Unsloth** : optimisations 2-5× plus rapides que PEFT vanilla.

## Cas d'usage typiques

- **Style transfer** : un modèle qui parle "votre" style (voix de marque, tone juridique).
- **Domaine spécialisé** : médical, légal, technique pointu.
- **Task-specific** : classification fine, extraction structurée d'un domaine particulier.
- **Multilingue ciblé** : améliorer un modèle sur une langue spécifique.

## Quand préférer RAG

RAG (Retrieval-Augmented Generation) reste meilleur pour les **connaissances volatiles** (docs qui changent souvent). Fine-tuning est meilleur pour **comportement et style**. Les deux se combinent bien.

## The classic fine-tuning problem

Fine-tuning a full model (full fine-tuning) requires:

- Memory = model size × 4-6 (weights + gradients + optimizer state).
- For a 7B: ~70 GB of VRAM. For a 70B: ~500 GB.
- Cost: several thousand dollars in compute per run.

Out of reach for 99% of teams.

## The LoRA idea

Instead of training all weights, **LoRA adds small adaptation matrices** (low rank) on certain layers (typically attention layers). These matrices have ~1% of the model's parameters.

```
Full model: W (7B params, frozen)
LoRA: W' = W + A·B  where A,B are small (~70M trainable params)
```

Consequences:

- **Memory**: ~10 GB to fine-tune a 7B.
- **Time**: a few hours on a consumer GPU (RTX 4090).
- **Output**: a ~50 MB `lora.safetensors` file to apply to the base model.

## QLoRA — even more efficient

QLoRA combines LoRA + base model 4-bit quantization. Result:

- **Fine-tune a 70B on a single A100 80GB** (vs 8× A100 in full FT).
- **Fine-tune a 7B on RTX 3090 24GB**.
- Quality comparable to full fine-tuning.

## Practical tools

- **PEFT** (Hugging Face): reference Python lib.
- **Axolotl**: wrapper framework with readable YAML configs.
- **MLX-LM**: for Mac M-series, native fine-tuning.
- **Unsloth**: 2-5× faster optimizations than vanilla PEFT.

## Typical use cases

- **Style transfer**: a model speaking "your" style (brand voice, legal tone).
- **Specialized domain**: medical, legal, deep technical.
- **Task-specific**: fine classification, domain-specific structured extraction.
- **Targeted multilingual**: improve a model on a specific language.

## When to prefer RAG

RAG (Retrieval-Augmented Generation) stays better for **volatile knowledge** (docs that change often). Fine-tuning is better for **behavior and style**. The two combine well.
